Compatibility with `numpy2`

**Describe the bug**
`pyopencl` cannot be built with numpy2

**To Reproduce**
Steps to reproduce the behavior:
1. `git clone https://github.com/inducer/pyopencl`
2. `python setup.py build`

**Expected behavior**
A clear and concise description of what you expected to happen.

**Environment (please complete the following information):**
 - OS: Linux
 - ICD Loader and version: 2.3.1-1
 - ICD and version: nvidia + PoCL
 - CPU/GPU: any Nvidia A5000 
  - Python version: 3.12
  - PyOpenCL version: 2024.1 (main branch)

**Additional context**
Python from mamba, build in a virtual-env (not conda-env)

```
~/workspace/pyopencl % python setup.py build
running build
running build_py
copying pyopencl/cache.py -> build/lib.linux-x86_64-cpython-312/pyopencl
copying pyopencl/ipython_ext.py -> build/lib.linux-x86_64-cpython-312/pyopencl
copying pyopencl/cltypes.py -> build/lib.linux-x86_64-cpython-312/pyopencl
copying pyopencl/tools.py -> build/lib.linux-x86_64-cpython-312/pyopencl
copying pyopencl/__init__.py -> build/lib.linux-x86_64-cpython-312/pyopencl
copying pyopencl/_cluda.py -> build/lib.linux-x86_64-cpython-312/pyopencl
copying pyopencl/_mymako.py -> build/lib.linux-x86_64-cpython-312/pyopencl
copying pyopencl/invoker.py -> build/lib.linux-x86_64-cpython-312/pyopencl
copying pyopencl/version.py -> build/lib.linux-x86_64-cpython-312/pyopencl
copying pyopencl/scan.py -> build/lib.linux-x86_64-cpython-312/pyopencl
copying pyopencl/bitonic_sort.py -> build/lib.linux-x86_64-cpython-312/pyopencl
copying pyopencl/reduction.py -> build/lib.linux-x86_64-cpython-312/pyopencl
copying pyopencl/elementwise.py -> build/lib.linux-x86_64-cpython-312/pyopencl
copying pyopencl/bitonic_sort_templates.py -> build/lib.linux-x86_64-cpython-312/pyopencl
copying pyopencl/algorithm.py -> build/lib.linux-x86_64-cpython-312/pyopencl
copying pyopencl/capture_call.py -> build/lib.linux-x86_64-cpython-312/pyopencl
copying pyopencl/clrandom.py -> build/lib.linux-x86_64-cpython-312/pyopencl
copying pyopencl/clmath.py -> build/lib.linux-x86_64-cpython-312/pyopencl
copying pyopencl/array.py -> build/lib.linux-x86_64-cpython-312/pyopencl
copying pyopencl/compyte/dtypes.py -> build/lib.linux-x86_64-cpython-312/pyopencl/compyte
copying pyopencl/compyte/array.py -> build/lib.linux-x86_64-cpython-312/pyopencl/compyte
copying pyopencl/characterize/__init__.py -> build/lib.linux-x86_64-cpython-312/pyopencl/characterize
copying pyopencl/characterize/performance.py -> build/lib.linux-x86_64-cpython-312/pyopencl/characterize
copying pyopencl/compyte/ndarray/gen_reduction.py -> build/lib.linux-x86_64-cpython-312/pyopencl/compyte/ndarray
copying pyopencl/compyte/ndarray/test_gpu_elemwise.py -> build/lib.linux-x86_64-cpython-312/pyopencl/compyte/ndarray
copying pyopencl/compyte/ndarray/test_gpu_ndarray.py -> build/lib.linux-x86_64-cpython-312/pyopencl/compyte/ndarray
copying pyopencl/compyte/ndarray/setup_opencl.py -> build/lib.linux-x86_64-cpython-312/pyopencl/compyte/ndarray
copying pyopencl/compyte/ndarray/gen_elemwise.py -> build/lib.linux-x86_64-cpython-312/pyopencl/compyte/ndarray
running egg_info
writing pyopencl.egg-info/PKG-INFO
writing dependency_links to pyopencl.egg-info/dependency_links.txt
writing requirements to pyopencl.egg-info/requires.txt
writing top-level names to pyopencl.egg-info/top_level.txt
reading manifest file 'pyopencl.egg-info/SOURCES.txt'
reading manifest template 'MANIFEST.in'
warning: no files found matching 'experiments/*.py'
warning: no previously-included files matching '*' found under directory '_skbuild'
adding license file 'LICENSE'
writing manifest file 'pyopencl.egg-info/SOURCES.txt'
/users/kieffer/.venv/py312/lib/python3.12/site-packages/setuptools/command/build_py.py:207: _Warning: Package 'pyopencl.cl' is absent from the `packages` configuration.
!!

        ********************************************************************************
        ############################
        # Package would be ignored #
        ############################
        Python recognizes 'pyopencl.cl' as an importable package[^1],
        but it is absent from setuptools' `packages` configuration.

        This leads to an ambiguous overall configuration. If you want to distribute this
        package, please make sure that 'pyopencl.cl' is explicitly added
        to the `packages` configuration field.

        Alternatively, you can also rely on setuptools' discovery methods
        (for example by using `find_namespace_packages(...)`/`find_namespace:`
        instead of `find_packages(...)`/`find:`).

        You can read more about "package discovery" on setuptools documentation page:

        - https://setuptools.pypa.io/en/latest/userguide/package_discovery.html

        If you don't want 'pyopencl.cl' to be distributed and are
        already explicitly excluding 'pyopencl.cl' via
        `find_namespace_packages(...)/find_namespace` or `find_packages(...)/find`,
        you can try to use `exclude_package_data`, or `include-package-data=False` in
        combination with a more fine grained `package-data` configuration.

        You can read more about "package data files" on setuptools documentation page:

        - https://setuptools.pypa.io/en/latest/userguide/datafiles.html


        [^1]: For Python, any directory (with suitable naming) can be imported,
              even if it does not contain any `.py` files.
              On the other hand, currently there is no concept of package data
              directory, all directories are treated like packages.
        ********************************************************************************

!!
  check.warn(importable)
copying pyopencl/cl/pyopencl-bessel-j-complex.cl -> build/lib.linux-x86_64-cpython-312/pyopencl/cl
copying pyopencl/cl/pyopencl-complex.h -> build/lib.linux-x86_64-cpython-312/pyopencl/cl
copying pyopencl/cl/pyopencl-hankel-complex.cl -> build/lib.linux-x86_64-cpython-312/pyopencl/cl
copying pyopencl/cl/pyopencl-ranluxcl.cl -> build/lib.linux-x86_64-cpython-312/pyopencl/cl
copying pyopencl/cl/pyopencl-random123/array.h -> build/lib.linux-x86_64-cpython-312/pyopencl/cl/pyopencl-random123
copying pyopencl/cl/pyopencl-random123/openclfeatures.h -> build/lib.linux-x86_64-cpython-312/pyopencl/cl/pyopencl-random123
copying pyopencl/cl/pyopencl-random123/philox.cl -> build/lib.linux-x86_64-cpython-312/pyopencl/cl/pyopencl-random123
copying pyopencl/cl/pyopencl-random123/threefry.cl -> build/lib.linux-x86_64-cpython-312/pyopencl/cl/pyopencl-random123
running build_ext
gcc -pthread -B /opt/mamba/envs/python_3.12/compiler_compat -fno-strict-overflow -Wall -O3 -DNDEBUG -fPIC -I/users/kieffer/.venv/py312/include -I/opt/mamba/envs/python_3.12/include/python3.12 -c /tmp/tmps_in1_fz.cpp -o tmp/tmps_in1_fz.o -std=gnu++14
gcc -pthread -B /opt/mamba/envs/python_3.12/compiler_compat -fno-strict-overflow -Wall -O3 -DNDEBUG -fPIC -I/users/kieffer/.venv/py312/include -I/opt/mamba/envs/python_3.12/include/python3.12 -c /tmp/tmpcnetwv0g.cpp -o tmp/tmpcnetwv0g.o -fvisibility=hidden
INFO: Disabling color, you really want to install colorlog.
Disabling color, you really want to install colorlog.
building 'pyopencl._cl' extension
gcc -pthread -B /opt/mamba/envs/python_3.12/compiler_compat -fno-strict-overflow -Wall -O3 -DNDEBUG -fPIC -DPYGPU_PACKAGE=pyopencl -DPYGPU_PYOPENCL=1 -DPYOPENCL_PRETEND_CL_VERSION=4112 -Ipybind11/include -I/users/kieffer/.venv/py312/lib/python3.12/site-packages/pybind11/include -I/users/kieffer/.venv/py312/lib/python3.12/site-packages/numpy/core/include -I/users/kieffer/.venv/py312/include -I/opt/mamba/envs/python_3.12/include/python3.12 -c -std=gnu++14 src/bitlog.cpp -o build/temp.linux-x86_64-cpython-312/src/bitlog.o -fvisibility=hidden -DVERSION_INFO=\"2024.1\" -fvisibility=hidden
gcc -pthread -B /opt/mamba/envs/python_3.12/compiler_compat -fno-strict-overflow -Wall -O3 -DNDEBUG -fPIC -DPYGPU_PACKAGE=pyopencl -DPYGPU_PYOPENCL=1 -DPYOPENCL_PRETEND_CL_VERSION=4112 -Ipybind11/include -I/users/kieffer/.venv/py312/lib/python3.12/site-packages/pybind11/include -I/users/kieffer/.venv/py312/lib/python3.12/site-packages/numpy/core/include -I/users/kieffer/.venv/py312/include -I/opt/mamba/envs/python_3.12/include/python3.12 -c -std=gnu++14 src/wrap_cl.cpp -o build/temp.linux-x86_64-cpython-312/src/wrap_cl.o -fvisibility=hidden -DVERSION_INFO=\"2024.1\" -fvisibility=hidden
In file included from src/wrap_cl.cpp:29:
src/wrap_cl.hpp:101:10: fatal error: numpy/arrayobject.h: No such file or directory
  101 | #include <numpy/arrayobject.h>
      |          ^~~~~~~~~~~~~~~~~~~~~
compilation terminated.
error: command '/usr/lib/ccache/gcc' failed with exit code 1
```

--- a/aksetup_helper.py
+++ b/aksetup_helper.py
@@ -43,14 +43,21 @@ def setup(*args, **kwargs):
 
 
 def get_numpy_incpath():
-    from os.path import join, dirname
+    from os.path import join, dirname, exists
     from importlib.util import find_spec
     origin = find_spec("numpy").origin
     if origin is None:
         raise RuntimeError("origin of numpy package not found")
 
     pathname = dirname(origin)
-    return join(pathname, "core", "include")
+    for p in [
+        join(pathname, "_core", "include"),  # numpy 2 onward
+        join(pathname, "core", "include"),  # numpy prior to 2
+    ]:
+        if exists(join(p, "numpy", "arrayobject.h")):
+            return p
+
+    raise RuntimeError("no valid path for numpy found")
 
 
 class NumpyExtension(Extension):

--- a/pyopencl/array.py
+++ b/pyopencl/array.py
@@ -56,6 +56,9 @@ else:
     _SVMPointer_or_nothing = ()
 
 
+_NUMPY_PRE_2 = np.__version__.startswith("1.")
+
+
 # {{{ _get_common_dtype
 
 _COMMON_DTYPE_CACHE: Dict[Tuple[Hashable, ...], np.dtype] = {}
@@ -123,7 +126,9 @@ def _get_common_dtype(obj1, obj2, queue):
     # Note that np.find_common_type, while appealing, won't be able to tell
     # the full story.
 
-    if not (o1_is_array and o2_is_array) and o1_is_integral and o2_is_integral:
+    if (_NUMPY_PRE_2
+            and not (o1_is_array and o2_is_array)
+            and o1_is_integral and o2_is_integral):
         if o1_is_array:
             obj1 = np.zeros(1, dtype=o1_dtype)
         if o2_is_array:
@@ -1413,7 +1418,7 @@ class Array:
     def __itruediv__(self, other):
         # raise an error if the result cannot be cast to self
         common_dtype = _get_truedivide_dtype(self, other, self.queue)
-        if not np.can_cast(common_dtype, self.dtype.type):
+        if not np.can_cast(common_dtype, self.dtype.type, "same_kind"):
             raise TypeError(
                 "Cannot cast {!r} to {!r}".format(self.dtype, common_dtype))
 

--- a/src/wrap_cl.hpp
+++ b/src/wrap_cl.hpp
@@ -139,6 +139,10 @@
 
 // {{{ macros and typedefs for wrappers
 
+#if NPY_ABI_VERSION < 0x02000000
+  #define PyDataType_ELSIZE(descr) ((descr)->elsize)
+#endif
+
 #if PY_VERSION_HEX >= 0x02050000
   typedef Py_ssize_t PYOPENCL_BUFFER_SIZE_T;
 #else
@@ -3477,11 +3481,12 @@ namespace pyopencl
     PYOPENCL_PARSE_WAIT_FOR;
     PYOPENCL_PARSE_NUMPY_ARRAY_SPEC;
 
-    npy_uintp size_in_bytes = tp_descr->elsize;
+    npy_uintp size_in_bytes = PyDataType_ELSIZE(tp_descr);
     for (npy_intp sdim: shape)
       size_in_bytes *= sdim;
 
     py::object result;
+    PyArrayObject *result_arr;
 
     cl_event evt;
     cl_int status_code;
@@ -3506,6 +3511,7 @@ namespace pyopencl
     event evt_handle(evt, false);
 
     std::unique_ptr<memory_map> map;
+
     try
     {
       result = py::object(py::reinterpret_steal<py::object>(PyArray_NewFromDescr(
@@ -3515,7 +3521,8 @@ namespace pyopencl
           strides.empty() ? nullptr : &strides.front(),
           mapped, ary_flags, /*obj*/nullptr)));
 
-      if (size_in_bytes != (npy_uintp) PyArray_NBYTES(result.ptr()))
+      result_arr = (PyArrayObject *) result.ptr();
+      if (size_in_bytes != (npy_uintp) PyArray_NBYTES(result_arr))
         throw pyopencl::error("enqueue_map_buffer", CL_INVALID_VALUE,
             "miscalculated numpy array size (not contiguous?)");
 
@@ -3529,7 +3536,7 @@ namespace pyopencl
     }
 
     py::object map_py(handle_from_new_ptr(map.release()));
-    PyArray_BASE(result.ptr()) = map_py.ptr();
+    PyArray_SetBaseObject(result_arr, map_py.ptr());
     Py_INCREF(map_py.ptr());
 
     return py::make_tuple(
@@ -3601,9 +3608,10 @@ namespace pyopencl
         shape.empty() ? nullptr : &shape.front(),
         strides.empty() ? nullptr : &strides.front(),
         mapped, ary_flags, /*obj*/nullptr));
+    PyArrayObject *result_arr = (PyArrayObject *) result.ptr();
 
     py::object map_py(handle_from_new_ptr(map.release()));
-    PyArray_BASE(result.ptr()) = map_py.ptr();
+    PyArray_SetBaseObject(result_arr, map_py.ptr());
     Py_INCREF(map_py.ptr());
 
     return py::make_tuple(
@@ -5762,13 +5770,13 @@ namespace pyopencl
         dims.push_back(py::cast<npy_intp>(it));
     }
 
-    NPY_ORDER order = PyArray_CORDER;
+    NPY_ORDER order = NPY_CORDER;
     PyArray_OrderConverter(order_py.ptr(), &order);
 
     int ary_flags = 0;
-    if (order == PyArray_FORTRANORDER)
+    if (order == NPY_FORTRANORDER)
       ary_flags |= NPY_FARRAY;
-    else if (order == PyArray_CORDER)
+    else if (order == NPY_CORDER)
       ary_flags |= NPY_CARRAY;
     else
       throw std::runtime_error("unrecognized order specifier");
@@ -5786,13 +5794,14 @@ namespace pyopencl
         &PyArray_Type, tp_descr,
         dims.size(), &dims.front(), /*strides*/ nullptr,
         host_ptr, ary_flags, /*obj*/nullptr));
+    PyArrayObject *result_arr = (PyArrayObject *) result.ptr();
 
-    if ((size_t) PyArray_NBYTES(result.ptr()) > mem_obj_size)
+    if ((size_t) PyArray_NBYTES(result_arr) > mem_obj_size)
       throw pyopencl::error("MemoryObject.get_host_array",
           CL_INVALID_VALUE,
           "Resulting array is larger than memory object.");
 
-    PyArray_BASE(result.ptr()) = mem_obj_py.ptr();
+    PyArray_SetBaseObject(result_arr, mem_obj_py.ptr());
     Py_INCREF(mem_obj_py.ptr());
 
     return result;

--- a/src/wrap_helpers.hpp
+++ b/src/wrap_helpers.hpp
@@ -132,13 +132,13 @@ namespace py = pybind11;
       COPY_PY_LIST(npy_intp, shape); \
     } \
     \
-    NPY_ORDER order = PyArray_CORDER; \
+    NPY_ORDER order = NPY_CORDER; \
     PyArray_OrderConverter(py_order.ptr(), &order); \
     \
     int ary_flags = 0; \
-    if (order == PyArray_FORTRANORDER) \
+    if (order == NPY_FORTRANORDER) \
       ary_flags |= NPY_FARRAY; \
-    else if (order == PyArray_CORDER) \
+    else if (order == NPY_CORDER) \
       ary_flags |= NPY_CARRAY; \
     else \
       throw std::runtime_error("unrecognized order specifier"); \

--- a/test/test_array.py
+++ b/test/test_array.py
@@ -683,7 +683,6 @@ def test_divide_inplace_scalar(ctx_factory):
     from itertools import product
 
     for dtype_a, dtype_s in product(dtypes, repeat=2):
-
         a = np.array([10, 20, 30, 40, 50, 60, 70, 80, 90, 100]).astype(dtype_a)
         s = dtype_s(40)
         a_gpu = cl_array.to_device(queue, a)
@@ -785,15 +784,23 @@ def test_bitwise(ctx_factory):
 
             assert (res_dev.get() == res).all()
 
-            res_dev = op(a_dev, s)
-            res = op(a, s)
+            try:
+                res = op(a, s)
+            except OverflowError:
+                pass
+            else:
+                res_dev = op(a_dev, s)
 
-            assert (res_dev.get() == res).all()
+                assert (res_dev.get() == res).all()
 
-            res_dev = op(s, b_dev)
-            res = op(s, b)
+            try:
+                res = op(s, b)
+            except OverflowError:
+                pass
+            else:
+                res_dev = op(s, b_dev)
 
-            assert (res_dev.get() == res).all()
+                assert (res_dev.get() == res).all()
 
         for op in [o.iand, o.ior, o.ixor]:
             res_dev = a_dev.copy()
@@ -801,17 +808,25 @@ def test_bitwise(ctx_factory):
             assert op_res is res_dev
 
             res = a.copy()
-            op(res, b)
+            try:
+                op(res, b)
+            except OverflowError:
+                pass
+            else:
 
-            assert (res_dev.get() == res).all()
+                assert (res_dev.get() == res).all()
 
-            res_dev = a_dev.copy()
-            op_res = op(res_dev, s)
-            assert op_res is res_dev
             res = a.copy()
-            op(res, s)
+            try:
+                op(res, s)
+            except OverflowError:
+                pass
+            else:
+                res_dev = a_dev.copy()
+                op_res = op(res_dev, s)
+                assert op_res is res_dev
 
-            assert (res_dev.get() == res).all()
+                assert (res_dev.get() == res).all()
 
         # Test unary ~
         res_dev = ~a_dev

